book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+primitive types, but have only one idiomatic way of doing so.
+
+In practice this means that builtin types include only primitives (int, text,
+bool, float), struct (for product), enum (for sum) and list (for repeating).
+
+An SQL row would translate to struct, and a relation would translate to a list
+of structs.
+
+I would also strive for the type system to be minimal - don't differentiate
+between tuples, objects and structs. Choose one and stick to it.
+
+**Type constraints** - constrain a type with a predicate. For example, have a
+type of int64 that are equal or greater than 10. Postgres
+[does support this](https://news.ycombinator.com/item?id=34835063). The primary
+value of using constrained types would not be validation (as it is used in
+linked article), but when matching the type.
+
+Say, for example, that we have a pipeline like this:
+
+```
+derive color = switch [x => 'red', true => 'green']
+derive is_red = switch [color == 'red' => true, color == 'green' => false]
+```
+
+It should be possible to infer that `color` is of type `text`, but only when
+equal to `'red'` or `'green'`. This means that the second switch covers all
+possible cases and `is_red` cannot be `null`.
+
+## Physical layout
+
+_Logical type_ is user-facing the notion of a type that is the building block of
+the type system.
+
+_Physical layout_ is the underlying memory layout of the data represented by a
+variable.
+
+In many programming languages, physical layout of a logical type is dependent on
+the target platform. Similarly, physical layout of a PRQL logical type is
+dependent on representation of that type in target STS.
+
+```
+PTS logical type  --->  STS logical type  ---> STS physical layout
+```
+
+Note that STS types do not have a single physical layout. Postgres has a logical
+type `anyelement`, which is a super type of any data type. It can be used as a
+function parameter type, but does not have a single physical layout so it cannot
+be used in a column declaration.
+
+For now, PRQL does not define a physical layout of any type. It is not needed
+since PRQL is not used for DDL (see section "Built-in primitives") or does not
+support raw access to underlying memory.
+
+As a consequence, results of a PRQL query cannot be robustly compared across
+DBMSs, since the physical layout of the result will vary.
+
+In the future, PRQL may define a common physical layout of types, probably using
+Apache Arrow.
+
+## Syntax
+
+```
+# built-in types
+type int
+type float
+type bool
+type text
+type char
+type null
+
+# users-defined types
+type my_int = int
+
+# sum type (union or enum)
+type number = int | float
+type scalar = number | bool | str | char
+
+# by default types are not nullable
+type my_nullable_int = int | null
+
+# user-defined enum
+type open
+type pending
+type closed
+type status = open | pending | closed
+
+# product type
+type my_struct = [id = int_nullable, name = str]
```

Comment:
> @snth 
> Other syntax that we have already firmly established is (square) brackets [] for lists. I think we should carry this over into the type system rather than introducing {} for this. Moreover, in terms of point 2. above, {} is quite commonly used for struct types in other languages such as Javascript, JSON, Python, ... .
> 
> Moreover, we have already established : as syntax for named parameters, as for example in the interp function definition above. So then if we view {} as a "struct constructor", then it would make sense that it takes named parameters with : and purely positional ones (for tuples) simply separated by ,. This has the nice consequence that the resulting syntax is very similar, if not identical to, JSON.

There is a clash between what we now call "lists" and what list is in this PR.

Our `select` and `derive` take something that has a fixed number elements, they can be of different types and they can be named. This definition collides with what this PR proposes under the name "struct". So effectively, this PR renames "lists" to "structs".

E.g.:
```
select [1.4, false, "foo"] # this is now named "struct"
```

It also proposes a new construct named "list" which acts as "a repeated type" - number of repetitions is unknown at compile time, all elements are of the same type and they cannot be named. Think of a column in a relation.

Now, there are two options for the syntax:
- (what this PR currently proposes) try to preserve current syntax: use `[]` for structs and `{}` for lists,
- (what you described) mimic JSON and basically all other languages: use `{}` for structs and `[]` for lists.

Reply:
> This definition collides with what this PR proposes under the name "struct". So effectively, this PR renames "lists" to "structs".
> 
> E.g.:
> 
> ```
> select [1.4, false, "foo"] # this is now named "struct"
> ```

I hadn't thought about it like that, but I think that's quite accurate.

IIUC most DBs name their List type as Array, so we could have 

I'm keen to keep the PRQL construct of `select [a, b, c]` be called something simple, like List (and not Struct), since it's a very simple type that everyone would encounter very early. 

Whereas the "repeated type" is more advanced and OK to call something technical.

Reply:
We'd then have lists and arrays. And lists would represent relation rows.

What about tuples and arrays? Is the name "tuple" approachable for non-programmer people? Would it be confusing that `[]` is used for tuples?

I think that tuple is very close to relational algebra and would be a good fit here.

Reply:
> I think that tuple is very close to relational algebra and would be a good fit here.

It's a bit foreign but I agree it's very close to relational algebra. Definitely fine in the typing docs; I'm less keen on having the opening example explaining PRQL be "Then we pass a _tuple_ of `[1.4, false, "foo"]` to `select`"...

It's also unfortunate that our `[]` syntax is not standard tuple syntax. Though this could give us an opportunity to make our parentheses syntax even more complicated! (no)

> And lists would represent relation rows.

I didn't have this in my mind — if anything the List would be closer to columns, or values. (Or maybe that's what you mean? Values in one row?)

Reply:
This naming gets confusing fast :D Let me clarify:

- lists (currently) or tuples (proposed) (what we express with `[]`) correspond to rows.
- arrays (proposed) correspond to columns.

Reply:
Very interesting! I had thought about the `select` case briefly before posting but I had only considered the case of a list of column names so I thought it was still covered by `lists / arrays`. I had forgotten that you can also include literal values and do renaming like you mentioned which brings you to structs as you pointed out, e.g.

    select [num = 1.4, truthy = false, bar = "foo", double_col = col * 2]

Having `[ ]` for structs / records is uncommon but not unprecedented. This is actually taking us pretty close to [Power Query M Formula Language](https://learn.microsoft.com/en-us/powerquery-m/) because M also uses `[]` for records and `{}` for lists. See for example this example from their [Quick Tour](https://learn.microsoft.com/en-us/powerquery-m/quick-tour-of-the-power-query-m-formula-language):

    let Orders = Table.FromRecords({  
        [OrderID = 1, CustomerID = 1, Item = "fishing rod", Price = 100.0],  
        [OrderID = 2, CustomerID = 1, Item = "1 lb. worms", Price = 5.0],  
        [OrderID = 3, CustomerID = 2, Item = "fishing net", Price = 25.0]}),  
        #"Capitalized Each Word" = Table.TransformColumns(Orders, {"Item", Text.Proper})  
    in  
        #"Capitalized Each Word"

M is actually underappreciated IMHO, probably because few programmers bother with Power Query and the people that use Power Query just want to get the data out and use the UI and aren't too interested in the underlying M code. 

M is actually pretty similar to PRQL in it's aims and use cases, namely it's a **pipelined, functional language for data munging**. I quite like it in its structure and approach. 

What I don't like about it is:
* it tends to be quite verbose
* the syntax for structs and lists is unconventional (and the wrong way around IMHO)

Given these (at least to me) new insights that we've got now, I think we should take a careful look at PRQL syntax again and perhaps try out some variations just to see what they look like.

Reply:
Woah, M is really similar to PRQL in some aspects, nice find.

Looking at the example you've posted, I see the unconventional aspect of `[]` for structs and `{}` for lists.

I'm now leaning to the "mimic JSON" syntax. An example:

```elm
from my_table
select {num = 1.4, truthy = false, bar = "foo", double_col = col * 2}
group {col_2} (derive a = sum double_col)
sort {num, truthy}
```

Reply:
> Would it be possible to further clarify what we mean by "correspond to {rows,columns}"?

Right now, this is the way I'd model our relations:
```
relation = array of tuples
tuple = a list of named types
```
From this I hope it's obvious what I mean when I say that a relation row is expressed as a tuple.

----

> For most people who aren't programmers, the squiggly brackets are barely used!

I'd say the same thing for square brackets.

Reply:
I agree with @max-sixty that we should pause for a while before a change like this (if we went ahead) because it's quite fundamental. I'm actually pro it and would also favour bringing things around to be more consistent with JSON type syntax but IMHO we should take care to try and think through as many of the repercussions as we can because we don't want to make changes like this to often.

So, in terms of current release numbers I would propose something like the following (numbers are just indicative and could be incremented appropriately depending on when the work gets done):
* 0.6.0: release imminent, pretty much in "feature freeze" mode
* 0.7.0: more QOL improvements, (settle syntax deliberations)
* 0.8.0: last release with old syntax, advertise upcoming syntax change
* 0.9.0 or 0.10.0: release new syntax

My main point is that given that it might be a fair amount of work to rewrite existing queries, people should be given adequate warning and they should still be given the opportunity to benefit from the next round or two of QOL improvements before having to rewrite their queries for the new syntax.

Maybe I'm too conservative but that's how I would go about it.

---
I agree that we should probably move this to the Github Discussions section since we seem to have stumbled onto a bigger topic than we anticipated.

Reply:
> tuple = a list of named types

OK, makes sense.

---

(Let's move the syntax conversation elsewhere, I'm muddying the discussion...)

Reply:
Let's continue the discussion here: https://github.com/PRQL/prql/issues/2124